Add WithVariable to MathematicalProgram

hongkai-dai opened the pull request:
This mimics the method `WithVar` and `WithIndet` in spotless, that allows us to add variables constructed elsewhere to a mathematical program. This feature will be used in the branch and bound, when we construct the optimization program in the root node of the tree, from the mixed integer optimization program.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7666)
<!-- Reviewable:end -->

hongkai-dai commented:
+@ericcousineau-tri for feature review please.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1pLRmwufoYO2ogcykM:bvhsv4m)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, pending discussion items.

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[solvers/mathematical_program.cc, line 152 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sK8WEjnpdTJe11lZU:-L1sK8WFLYhK9oJq8OZc:b-7r4s38) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L152)):*
> ```cpp
> }
> 
> void MathematicalProgram::WithVariables(
> ```

Having the name `With{X}` implies that there's some sort of scoping to this statement (at least coming from Python-land).

Would you be amenable to renaming these to `Add{X}`? (renaming internal methods if they get in the way?)

---

*[solvers/mathematical_program.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sKQWZ0AM073ak_FZZ:-L1sKQWZ0AM073ak_FZ_:br2te0z) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L154)):*
> ```cpp
> void MathematicalProgram::WithVariables(
>     const Eigen::Ref<const VectorXDecisionVariable>& variables) {
>   const int num_existing_vars = num_vars();
> ```

Also, I do prefer the shorter name `Variables`; however, other portions of the code use `DecisionVariables`.
Should these be made consistent? (either changing the new or the old?)

---

*[solvers/mathematical_program.cc, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sJbF5ucIzXMTWAqBe:-L1sJbF7fvhTzKCFClm-:b-leut3u) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L158)):*
> ```cpp
>     if (variables(i).is_dummy()) {
>       std::ostringstream oss;
>       oss << "variables(" << i << ") is dummy.\n";
> ```

BTW For things like this, have you considered using `fmt::format("variables ({}) is dummy", i)`?

Also, this seems like a good use case for having a general purpose `DRAKE_THROW_UNLESS_FMT(...)` - but maybe less verbose.

Would you like me to make this? I'd need to check with Jeremy and see if he's OK with it.

---

*[solvers/test/mathematical_program_test.cc, line 2110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sL-nY4w2jypf9P4gK:-L1sL-nY4w2jypf9P4gL:beix70q) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/test/mathematical_program_test.cc#L2110)):*
> ```cpp
>                runtime_error);
>   EXPECT_THROW(prog_.AddLorentzConeConstraint(
>                    2 * x_(0),
> ```

Would it be possible to exclude the external whitespace formatting changes from this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1sL60O4XelI47bRtPi:b-2l0s3s)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[solvers/mathematical_program.h, line 518 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNVBYWt_LU5aOGI94:-L1sNVBYWt_LU5aOGI95:bvydw34) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L518)):*
> ```objc
>   }
> 
>   /**
> ```

BTW consistency Some comments are `/** Text...`, others are `/**\n * Text....`
Could these be made consistent?

---

*[solvers/mathematical_program.h, line 519 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sMyt9dX0ywyxH2v8i:-L1sMyt9dX0ywyxH2v8j:b-6qlz4d) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L519)):*
> ```objc
> 
>   /**
>    * Append some new variables to the end of the existing variables.
> ```

BTW grammar nit Should be `Appends`

---

*[solvers/mathematical_program.h, line 520 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sN-oMbv3IJ7N6k3Mu:-L1sN-oMbv3IJ7N6k3Mv:brnthrm) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L520)):*
> ```objc
>   /**
>    * Append some new variables to the end of the existing variables.
>    * @param variables The newly added variables.
> ```

BTW grammar nit `some` sounds a bit vagues. Perhaps just remove it?

---

*[solvers/mathematical_program.h, line 521 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sN3P9EdGKyYjoscIL:-L1sN3P9EdGKyYjoscIM:b50tlug) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L521)):*
> ```objc
>    * Append some new variables to the end of the existing variables.
>    * @param variables The newly added variables.
>    * @pre variables should not intersect with the existing variables in the
> ```

BTW Should ```variables``` be in backticks for `@pre`?

---

*[solvers/mathematical_program.h, line 523 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNb4gLBC5k1vv4nwQ:-L1sNb4gLBC5k1vv4nwR:bvku1r4) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L523)):*
> ```objc
>    * @pre variables should not intersect with the existing variables in the
>    * optimization program.
>    * @pre each entry in `variables` should not be a dummy variable.
> ```

BTW grammar nit Capitalization, here and in other places.

---

*[solvers/mathematical_program.h, line 524 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNwL7NPRzV-RrnzBS:-L1sNwL84G9c-GDWu-F9:bi1hzbs) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L524)):*
> ```objc
>    * optimization program.
>    * @pre each entry in `variables` should not be a dummy variable.
>    * @throw runtime_error if the preconditions are not satisfied.
> ```

FYI I like this!
(Should it be added to other sites?)

---

*[solvers/mathematical_program.h, line 704 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNf4KfZqlFB9E6r_8:-L1sNf4KfZqlFB9E6r_9:b-epvaz5) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L704)):*
> ```objc
>                                          const std::string& name = "X");
> 
>   /** Adds indeterminates to this MathematicalProgram.
> ```

BTW `to this MathematicalProgram` seems a tad redundant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[solvers/mathematical_program.cc, line 152 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sK8WEjnpdTJe11lZU:-L1t4ESplJeRvvYr5pIa:b3skfpg) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L152)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Having the name `With{X}` implies that there's some sort of scoping to this statement (at least coming from Python-land).

Would you be amenable to renaming these to `Add{X}`? (renaming internal methods if they get in the way?)
</blockquote></details>

Done. I changed to `Add{X}`.

---

*[solvers/mathematical_program.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sKQWZ0AM073ak_FZZ:-L1svyRkUC3-NSZXVEK3:bze6lgz) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L154)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Also, I do prefer the shorter name `Variables`; however, other portions of the code use `DecisionVariables`.
Should these be made consistent? (either changing the new or the old?)
</blockquote></details>

Done. I changed the new code to use `decision_vars`.

---

*[solvers/mathematical_program.cc, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sJbF5ucIzXMTWAqBe:-L1t4Hn4mIv8ETAZ6gU0:bxzbtj6) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L158)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW For things like this, have you considered using `fmt::format("variables ({}) is dummy", i)`?

Also, this seems like a good use case for having a general purpose `DRAKE_THROW_UNLESS_FMT(...)` - but maybe less verbose.

Would you like me to make this? I'd need to check with Jeremy and see if he's OK with it.
</blockquote></details>

Done. I changed to use `fmt::format`, this is a lot less code. Thanks!

---

*[solvers/mathematical_program.h, line 518 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNVBYWt_LU5aOGI94:-L1swQoHZLdGCi4Eb7OF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L518)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW consistency Some comments are `/** Text...`, others are `/**\n * Text....`
Could these be made consistent?
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 519 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sMyt9dX0ywyxH2v8i:-L1swSHn1JLr_jhTQymU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L519)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW grammar nit Should be `Appends`
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 520 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sN-oMbv3IJ7N6k3Mu:-L1swVdPMf-oLgjcqlvn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L520)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW grammar nit `some` sounds a bit vagues. Perhaps just remove it?
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 521 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sN3P9EdGKyYjoscIL:-L1swZSoozv1xDPYcQDC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L521)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Should ```variables``` be in backticks for `@pre`?
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 523 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNb4gLBC5k1vv4nwQ:-L1swdZdTmv3woEjR_1Q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L523)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW grammar nit Capitalization, here and in other places.
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 704 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sNf4KfZqlFB9E6r_8:-L1swrOttfBuhlu7LH0z:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.h#L704)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW `to this MathematicalProgram` seems a tad redundant.
</blockquote></details>

Done.

---

*[solvers/test/mathematical_program_test.cc, line 2110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sL-nY4w2jypf9P4gK:-L1sx1mM29_TUt2gG1BY:bg7wr9b) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/test/mathematical_program_test.cc#L2110)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Would it be possible to exclude the external whitespace formatting changes from this PR?
</blockquote></details>

If it doesn't hurt, I think it is OK to keep the formatting changes? I do not know how to undo the changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@seancurtis-tri for platform review please, since Tuesday is almost over at Cambridge.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1t4dWnm0tm_b8EKT54:b-tiau3d)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-@SeanCurtis-TRI for platform review. +@ggould-tri for platform review please.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1tOPipmRYAozPLOwLa:biirv9y)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
A few items below; I'll take another look tomorrow.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[solvers/mathematical_program.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sKQWZ0AM073ak_FZZ:-L1tg1WOUZyF4nQuYV0W:bxntqje) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L154)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. I changed the new code to use `decision_vars`.
</blockquote></details>

I think maybe this made things worse.  Now the argument is `variables`, its contents will be added to `decision_variables_`, the `num_existing_decision_vars` initialized from `num_vars()` blurs the distinction further, and `variables` can collide with `decision_variable_index_` or `indeterminates_index_` (which is, oddly, plural) .

If we have some terminological confusion over "decision variable" versus "variable" (versus "indeterminate"), can it be at least documented?  Or could we just decide on one name or the other?

Alternatively I am misunderstanding all of this and there is a good reason for the names; if so, it begs for a comment.

---

*[solvers/mathematical_program.h, line 520 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1thgLyc3iKbyQT9-Xy:-L1thgLyc3iKbyQT9-Xz:bpi40sc) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/mathematical_program.h#L520)):*
> ```objc
>   /** Appends new variables to the end of the existing variables.
>    * @param variables The newly added variables.
>    * @pre `variables` should not intersect with the existing variables in the
> ```

intersect with the existing variables *or indeterminates*, I think?

---

*[solvers/test/mathematical_program_test.cc, line 389 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1ti-SH08HoEml4adUE:-L1ti-SH08HoEml4adUF:b-g8f1mk) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L389)):*
> ```cpp
> 
> GTEST_TEST(testAddVariables, AddVariables1) {
>   // Call WithVariable on an empty program.
> ```

You renamed `WithVariable` so this comment needs to update.

---

*[solvers/test/mathematical_program_test.cc, line 410 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1ti9ndBvKL1VyXfaCo:-L1ti9ndBvKL1VyXfaCp:b-g8f1mk) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L410)):*
> ```cpp
> 
> GTEST_TEST(testAddVariables, AddVariable2) {
>   // Call WithVariable on a program that has some existing variables.
> ```

You renamed `WithVariable` so this comment needs to update.

---

*[solvers/test/mathematical_program_test.cc, line 572 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1tiXVa2ciJbef7-cNG:-L1tiXVa2ciJbef7-cNH:b-qv5h3h) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L572)):*
> ```cpp
>   EXPECT_THROW(prog.AddIndeterminates(VectorIndeterminate<2>(x0, dummy)),
>                std::runtime_error);
> }
> ```

Since the purpose of this PR is to allow reusing variables between different `MathematicalProgram`s, shouldn't you have a test case that does that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1tixN7ls2e7wjTYAah:b-opehi2)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[solvers/mathematical_program.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sKQWZ0AM073ak_FZZ:-L1uBrQmR_sXBfQ20Qhi:b-adgjdu) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/mathematical_program.cc#L154)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

I think maybe this made things worse.  Now the argument is `variables`, its contents will be added to `decision_variables_`, the `num_existing_decision_vars` initialized from `num_vars()` blurs the distinction further, and `variables` can collide with `decision_variable_index_` or `indeterminates_index_` (which is, oddly, plural) .

If we have some terminological confusion over "decision variable" versus "variable" (versus "indeterminate"), can it be at least documented?  Or could we just decide on one name or the other?

Alternatively I am misunderstanding all of this and there is a good reason for the names; if so, it begs for a comment.
</blockquote></details>

Done. I changed the function name to `AddDecisionVariables`, and variable name to `decision_variables`.

---

*[solvers/mathematical_program.h, line 520 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1thgLyc3iKbyQT9-Xy:-L1uBnLn6ycVH19do3U2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/mathematical_program.h#L520)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

intersect with the existing variables *or indeterminates*, I think?
</blockquote></details>

Done.

---

*[solvers/test/mathematical_program_test.cc, line 389 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1ti-SH08HoEml4adUE:-L1uBoXftZRsL1jNPV8V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L389)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

You renamed `WithVariable` so this comment needs to update.
</blockquote></details>

Done.

---

*[solvers/test/mathematical_program_test.cc, line 410 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1ti9ndBvKL1VyXfaCo:-L1uBp4mILvlrJtK1X2q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L410)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

You renamed `WithVariable` so this comment needs to update.
</blockquote></details>

Done.

---

*[solvers/test/mathematical_program_test.cc, line 572 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1tiXVa2ciJbef7-cNG:-L1uCgatf5VCHx_JE5-v:b-g1p55l) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L572)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Since the purpose of this PR is to allow reusing variables between different `MathematicalProgram`s, shouldn't you have a test case that does that?
</blockquote></details>

We could reuse variables between different `MathematicalProgram`s, but that is not the primary purpose for this PR. I introduced this PR so that we can construct the variable elsewhere, instead of from a `MathematicalProgram`, and then add these variables to a `MathematicalProgram` object.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[solvers/test/mathematical_program_test.cc, line 572 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1tiXVa2ciJbef7-cNG:-L1wSZemY0g4Pqmhh9u6:b90nwmy) ([raw file](https://github.com/robotlocomotion/drake/blob/bcdd0058e17e55847e7818ac6d909a08f5a63bd6/solvers/test/mathematical_program_test.cc#L572)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

We could reuse variables between different `MathematicalProgram`s, but that is not the primary purpose for this PR. I introduced this PR so that we can construct the variable elsewhere, instead of from a `MathematicalProgram`, and then add these variables to a `MathematicalProgram` object.
</blockquote></details>

Ah, okay, I guess I misunderstood.  Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1wSdos8J7NjBaebWNN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[solvers/mathematical_program.cc, line 16 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1wVz36TnVQ49-0fdxf:-L1wVz37y2bbf-fwLTyI:b-bzm5e2) ([raw file](https://github.com/robotlocomotion/drake/blob/d1ecb3549de4020cc18f0b506ce59a590d226f76/solvers/mathematical_program.cc#L16)):*
> ```cpp
> #include <vector>
> 
> #include <spdlog/fmt/ostr.h>
> ```

I think this may be an indirect include. Can you try `#include <fmt/format.h>` for a more direct inclusion?

---

*[solvers/mathematical_program.cc, line 258 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1wWoIA9T5l3mDeN0KM:-L1wWoIBINGAU8t685Mj:b-17j1xq) ([raw file](https://github.com/robotlocomotion/drake/blob/d1ecb3549de4020cc18f0b506ce59a590d226f76/solvers/mathematical_program.cc#L258)):*
> ```cpp
>     if (new_indeterminates(i).is_dummy()) {
>       throw std::runtime_error(
>           fmt::format("new_indeterminates({}) is dummy.", i));
> ```

BTW Should this be reworded to say `new_indeterminates({}) cannot be a dummy variable.`?

---

*[solvers/test/mathematical_program_test.cc, line 2110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1sL-nY4w2jypf9P4gK:-L1wW_y2s6ouSiKadfbE:busotgq) ([raw file](https://github.com/robotlocomotion/drake/blob/06400ee07cfb76143f01c7f5256f5c7a050c9827/solvers/test/mathematical_program_test.cc#L2110)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

If it doesn't hurt, I think it is OK to keep the formatting changes? I do not know how to undo the changes.
</blockquote></details>

Yup, that's fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[solvers/mathematical_program.cc, line 16 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1wVz36TnVQ49-0fdxf:-L1wlZUz-A7fahSSEBXc:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/d1ecb3549de4020cc18f0b506ce59a590d226f76/solvers/mathematical_program.cc#L16)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

I think this may be an indirect include. Can you try `#include <fmt/format.h>` for a more direct inclusion?
</blockquote></details>

Done. Thanks!

---

*[solvers/mathematical_program.cc, line 258 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1wWoIA9T5l3mDeN0KM:-L1wlg07_g69robNUuIu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d1ecb3549de4020cc18f0b506ce59a590d226f76/solvers/mathematical_program.cc#L258)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Should this be reworded to say `new_indeterminates({}) cannot be a dummy variable.`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[solvers/mathematical_program.cc, line 258 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1wWoIA9T5l3mDeN0KM:-L1wvNcDYLYL-ZYeS1cX:b-isb40w) ([raw file](https://github.com/robotlocomotion/drake/blob/d1ecb3549de4020cc18f0b506ce59a590d226f76/solvers/mathematical_program.cc#L258)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

BTW Could you make this change to `new_indeterminates(...)...` as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[solvers/mathematical_program.cc, line 258 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-L1wWoIA9T5l3mDeN0KM:-L1x6aA3lsL4wATzDpe7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d1ecb3549de4020cc18f0b506ce59a590d226f76/solvers/mathematical_program.cc#L258)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Could you make this change to `new_indeterminates(...)...` as well?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7666#-:-L1xGyuQ6_hE4JmZz0Ze:b-l70wzf)*
<!-- Sent from Reviewable.io -->
